src/inference/index.ts

```diff
@@ -0,0 +1,552 @@
+/* eslint-disable @typescript-eslint/no-use-before-define */
+/* eslint-disable no-param-reassign */
+import { SchemaTypeID, PathStatistics } from '../models';
+
+interface SchemaObject {
+  [key: string]: SchemaType;
+}
+
+export class SchemaType {
+  /**
+   * Unique type ID that can be used to discriminate between different Schema
+   * Types like NumberType, StringType, ObjectType
+   */
+  public type: SchemaTypeID;
+
+  /**
+   * A user-defined identifier that is used to label every node of the schema.
+   * In the context of an Algolia Record, the objectID could be used as a marker.
+   *
+   * The markers are useless during the conversion of a JSON object but can be very
+   * useful during the combination of the models, as models that are different can
+   * then show different markers, which the user can then use to find what caused
+   * the difference.
+   */
+  public marker?: string;
+
+  /**
+   * A simple counter that counts how many times this part of the model was combined.
+   * This is useful to measure how many times an attribute is Missing, and compare it to
+   * how many times the parent object is present, for instance.
+   */
+  public counter: number;
+
+  public constructor(counter = 1, marker?: string) {
+    this.counter = counter;
+    this.marker = marker;
+    this.type = 'unknownType';
+  }
+
+  /**
+   * Generic method to merge two SchemaType into a single model of the correct
+   * type, and that can be overriden when a more advanced logic is needed
+   * (e.g. for Object, Arrays, etc.).
```

Comment:
it seems that you can both do `type1.combine(type2)` or `type2.combine(type1)`.

assuming that `type1` and `type2` are different kind of types and that only `type2` has overrided the `combine` method, wouldn't that break that "reversibility" property?

Reply:
If `type1 !== type2`, the behaviour is to transform it into a `UnionType<T1, T2>`, and this is the first check that is done in overriding `combine` methods.

It might be a good thing to document this requirement though 👍